File: hartshorn-commands/src/main/java/org/dockbox/hartshorn/commands/CommandSource.java
/*
 * Copyright 2019-2023 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.dockbox.hartshorn.commands;

import org.dockbox.hartshorn.context.ContextCarrier;
import org.dockbox.hartshorn.i18n.MessageReceiver;
import org.dockbox.hartshorn.util.Subject;

/**
 * Represents a single {@link Subject} capable of executing commands.
 */
public interface CommandSource extends MessageReceiver, ContextCarrier {

    /**
     * Executes the given raw command as the {@link Subject}
     * represented by this source.
     *
     * @param command The raw command
     */
    default void execute(String command) throws ParsingException {
        CommandGateway gateway = this.applicationContext().get(CommandGateway.class);
        gateway.accept(this, command);
    }
}

File: core/src/main/kotlin/org/dockbox/darwin/core/objects/targets/CommandSource.kt
package org.dockbox.darwin.core.objects.targets

import org.dockbox.darwin.core.i18n.I18N
import org.dockbox.darwin.core.text.Text

interface CommandSource : MessageReceiver {

    fun execute(command: String)

    object None : CommandSource {
        override fun execute(command: String) {
            throw UnsupportedOperationException("Attempted to execute command without source")
        }

        override fun send(text: I18N) {
            throw UnsupportedOperationException("Attempted to send message without source")
        }

        override fun send(text: Text) {
            throw UnsupportedOperationException("Attempted to send message without source")
        }

        override fun send(text: CharSequence) {
            throw UnsupportedOperationException("Attempted to send message without source")
        }

        override fun sendWithPrefix(text: I18N) {
            throw UnsupportedOperationException("Attempted to send message without source")
        }

        override fun sendWithPrefix(text: Text) {
            throw UnsupportedOperationException("Attempted to send message without source")
        }

        override fun sendWithPrefix(text: CharSequence) {
            throw UnsupportedOperationException("Attempted to send message without source")
        }
    }

}

